File: leet-code/622. Design Cirular Queue/DesignCircularQueue622_2.java
public class DesignCircularQueue622_2 {
    private final int[] circularQ;
    private final int maxSize;
    private int length = 0;
    private int front = 0;
    // 최초로 circularQ 에 요소를 넣을때
    // front 와 rear 가 같은 인덱스를 바라볼 수 있도록 한다
    private int rear = -1;

    public DesignCircularQueue622_2(int k) {
        circularQ = new int[k];
        maxSize = k;
    }

    public boolean enQueue(int value) {
        if (isFull()) {
            return false;
        }

        rear = (rear + 1) % maxSize;
        circularQ[rear] = value;
        length += 1;
        return true;
    }

    public boolean deQueue() {
        if (isEmpty()) {
            return false;
        }

        front = (front + 1) % maxSize;
        length -= 1;
        return true;
    }

    public int Front() {
        return isEmpty() ? -1 : circularQ[front];
    }

    public int Rear() {
        return isEmpty() ? -1 : circularQ[rear];
    }

    public boolean isEmpty() {
        return length == 0;
    }

    public boolean isFull() {
        return length == maxSize;
    }
}

File: leet-code/622. Design Cirular Queue/DesignCircularQueue622_2.kt
package com.example

class DesignCircularQueue622_2(k: Int) {
    private var front: Node? = null
    private var rear: Node? = null
    private val maxSize = k
    private var curSize = 0

    fun enQueue(value: Int): Boolean {
        if (isFull()) {
            return false
        }
        val oldRear = rear
        val newRear = Node(value)
        oldRear?.next = newRear
        rear = newRear
        if (isEmpty()) {
            front = rear
        }
        curSize += 1
        return true
    }

    fun deQueue(): Boolean {
        if (isEmpty()) {
            return false
        }
        val next = front!!.next
        front = next
        rear!!.next = next
        curSize -= 1
        return true
    }

    fun Front(): Int {
        return if (isEmpty()) -1 else front!!.value
    }

    fun Rear(): Int {
        return if (isEmpty()) -1 else rear!!.value
    }

    fun isEmpty(): Boolean {
        return curSize == 0
    }

    fun isFull(): Boolean {
        return curSize == maxSize
    }

    private class Node(k: Int) {
        val value = k
        var next: Node? = null
    }
}
